core/commands/object/diff.go

```diff
@@ -0,0 +1,103 @@
+package objectcmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	dagutils "github.com/ipfs/go-ipfs/merkledag/utils"
+	path "github.com/ipfs/go-ipfs/path"
+)
+
+var ObjectDiffCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "takes a diff of the two given objects",
+		ShortDescription: `
+ipfs object diff is a command used to show the differences between
+two ipfs objects.
+
+Example:
+
+   $ ls foo
+   bar baz/ giraffe
+   $ ipfs add -r foo
+   ...
+   added QmegHcnrPgMwC7tBiMxChD54fgQMBUecNw9nE9UUU4x1bz foo
+   $ OBJ_A=QmegHcnrPgMwC7tBiMxChD54fgQMBUecNw9nE9UUU4x1bz
+   $ echo "different content" > foo/bar
+   $ ipfs add -r foo
+   ...
+   added QmcmRptkSPWhptCttgHg27QNDmnV33wAJyUkCnAvqD3eCD foo
+   $ OBJ_B=QmcmRptkSPWhptCttgHg27QNDmnV33wAJyUkCnAvqD3eCD
+   $ ipfs object diff $OBJ_A $OBJ_B
+   changed "bar" from QmNgd5cz2jNftnAHBhcRUGdtiaMzb5Rhjqd4etondHHST8 to QmRfFVsjSXkhFxrfWnLpMae2M4GBVsry6VAuYYcji5MiZb
```

Comment:
it may be nice to make this a "more verbose" or "humanized" version of the output. and have something more diffstat like:

```
+ <added-link-name> <added-link-value>
- <removed-link-name> <removed-link-property>
~ <changed-link-name> <old-value> <new-value>
```
(or make this o/ an option)

---

another option (and relevant in light of IPLD) is to use json-diff or cbor-diff tooling.